ibis/bigquery/client.py

```diff
@@ -0,0 +1,207 @@
+import re
+
+import ibis
+import ibis.expr.types as ir
+import ibis.expr.datatypes as dt
+from ibis.bigquery import compiler as comp
+import toolz
+import google.datalab as gd
+import google.datalab.bigquery as bq
+from ibis.client import (Database, SQLClient)
+
+
+_fully_qualified_re = re.compile('(.*)\.(.*)|(.*)')
+
+
+# pip install git+https://github.com/googledatalab/pydatalab
+def _parse_bq_maybe_qualified_re(name):
+    (dataset, table0, table1) = _fully_qualified_re.match(name).groups()
+    table = table0 or table1
+    return (dataset, table)
+
+
+def _ensure_split(table_id, dataset_id):
+    (_dataset_id, table_id) = _parse_bq_maybe_qualified_re(table_id)
+    assert (dataset_id == _dataset_id) or not (dataset_id and _dataset_id)
+    dataset_id = dataset_id or _dataset_id
+    return (table_id, dataset_id)
+
+
+def _bq_table_obj_to_string(table):
+    (_, dataset_id, table_id, _) = table.name
+    return dataset_id + '.' + table_id
+
+
+def _bq_make_context(project_id=None):
+    context = gd.Context.default()
+    if context.project_id != project_id:
+        context.set_project_id(project_id)
+    return context
+
+
+@toolz.memoize
+def _bq_get_context(project_id):
```

Comment:
Do we need both `_bq_get_context` and `_bq_make_context`?

Reply:
making a context seems to be very costly in terms of time, so I expose a way to make one if you really want to but the default should probably be to just use get